src/impl/format.ts

```diff
@@ -77,87 +91,109 @@ export function format(documentText: string, range: Range | undefined, options:
 	}
 
 	while (firstToken !== SyntaxKind.EOF) {
+
 		let firstTokenEnd = scanner.getTokenOffset() + scanner.getTokenLength() + formatTextStart;
 		let secondToken = scanNext();
-
 		let replaceContent = '';
 		let needsLineBreak = false;
-		while (!lineBreak && (secondToken === SyntaxKind.LineCommentTrivia || secondToken === SyntaxKind.BlockCommentTrivia)) {
-			// comments on the same line: keep them on the same line, but ignore them otherwise
+
+		while (numberLineBreaks === 0 && (secondToken === SyntaxKind.LineCommentTrivia || secondToken === SyntaxKind.BlockCommentTrivia)) {
 			let commentTokenStart = scanner.getTokenOffset() + formatTextStart;
 			addEdit(' ', firstTokenEnd, commentTokenStart);
 			firstTokenEnd = scanner.getTokenOffset() + scanner.getTokenLength() + formatTextStart;
 			needsLineBreak = secondToken === SyntaxKind.LineCommentTrivia;
-			replaceContent = needsLineBreak ? newLineAndIndent() : '';
+			replaceContent = needsLineBreak ? newLinesAndIndent() : '';
 			secondToken = scanNext();
 		}
 
 		if (secondToken === SyntaxKind.CloseBraceToken) {
-			if (firstToken !== SyntaxKind.OpenBraceToken) {
-				indentLevel--;
-				replaceContent = newLineAndIndent();
+			if (firstToken !== SyntaxKind.OpenBraceToken) { indentLevel--; };
+
+			if (options.keepLines && numberLineBreaks > 0 || !options.keepLines && firstToken !== SyntaxKind.OpenBraceToken) {
+				replaceContent = newLinesAndIndent();
+			} else if (options.keepLines) {
+				replaceContent = ' ';
 			}
 		} else if (secondToken === SyntaxKind.CloseBracketToken) {
-			if (firstToken !== SyntaxKind.OpenBracketToken) {
-				indentLevel--;
-				replaceContent = newLineAndIndent();
+			if (firstToken !== SyntaxKind.OpenBracketToken) { indentLevel--; };
+
+			if (options.keepLines && numberLineBreaks > 0 || !options.keepLines && firstToken !== SyntaxKind.OpenBracketToken) {
+				replaceContent = newLinesAndIndent();
+			} else if (options.keepLines) {
+				replaceContent = ' ';
 			}
 		} else {
 			switch (firstToken) {
 				case SyntaxKind.OpenBracketToken:
 				case SyntaxKind.OpenBraceToken:
 					indentLevel++;
-					replaceContent = newLineAndIndent();
+					if (options.keepLines && numberLineBreaks > 0 || !options.keepLines) {
+						replaceContent = newLinesAndIndent();
+					} else if (options.keepLines) {
```

Comment:
I think at that point, `options.keepLines` is always true -> can remove the check